GetStream/stream-chat-net
# feat: small adjustments, plus DI package

## CLA

- [x] The code changes follow best practices
- [x] Code changes are tested (add some information if not applicable)

## Description of the pull request

Hi! I just opened this PR because I applied for the SDK Dev role and thought I could add some nice touches. 🙂 Quickly threw this together last night.

> :exclamation: Important: this pull request is **not** ready to merge in for a few reasons, I'll explain in the end.

The changes I made:
- Added Dependency Injection library which can be very easily used in ASP.NET Core web applications. A lot of other libraries does that (Autofac, AutoMapper etc. etc.), and it's comfortable for the user.
- Added GitHub Actions integration flow for the new Dependency Injection library. I could not add one for the "normal" library because the tests rely on valid API + Secret keys. Those tests needs to be refactored to not rely on those.
- Added the new feature to the README. Also recommended people to use a single instance of `Client`.
- Removed all default constructors, they aren't needed
- In fact, I was able to remove 174 lines from the code. I've refactored `RestClient` a little bit, so that it's a bit shorter and nicer to read
- Created multiple different constructors to `Client`. If it was a private a repository, I wouldn't bother with it, but for public libraries, it's just a nice touch to not have nullable parameters
- ❗ Most importantly: HttpClient must be a static/singleton. Everytime you instantiate a HttpClient, it creates a new socket. Yes, it was disposed properly, but it's still an overhead to create and close a new socket every time you invoke an API. It is Microsoft's recommendation to re-use the same client. I highlighted this in the README as well. See [here](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=net-5.0#remarks). Actually, this might be quite a decent performance improvement.
- Added ability to pass in your own HttpClient. Some people use proxies, some other people likes to attach their own handlers (for logging purposes, for example), so it's more comfortable for them. Additionally, it's easier to unit test the Client if we pass in a mocked `HttpClient`. Actually, it's impossible otherwise. 😬 
- Because of refactoring HttpClient, we needed to refactor Timeout handling a little bit as well using CancellationTokens.
- Disposing `HttpResponseMessage` in `RestResponse`
- `RestRequest.Headers` and `RestRequest.QueryParameters` are Dictionary instead of IEnumerable. In general, when you return data, if it was already evaluated (such as Dictionary), it makes more sense to return it, instead of a lazily evaluated array (IEnumerable). For example, the person can call `dictionary.Count` property instead of going through and counting all the elements which `ienumerable.Count()` does.
- `stream-chat-net.csproj` => netcoreapp target framework isn't needed, that is for applications. For libraries, `netstandard` should be used.

Testing
- I wasn't able to run the tests as I do not have api keys.

Now, as to why isn't it ready:
- The tests need to be refactored so that it doesn't rely on valid API keys. Those _should_ stay as well, but they need to be integration tests instead of unit tests.
- Most importantly: I would put the Dependency Injection library into a separate GitHub repo. That's because it's dependent on .NET Standard and the classic .NET Framework 4.* isn't supported. It's was super funky to separate the folders/projects so that I can actually run those tests. I really wouldn't do that honestly, it's much nicer to just have a separate repository with a different workflow.

Future ideas:
- Async methods must end with `Async` postfix as a general rule in C#. Might be a good idea to slowly deprecate old methods and encourage people to use the new Async methods. See [here](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap?redirectedfrom=MSDN) for reference.
- Adding docs to the method signatures.
- Rewriting the tests a little bit so that we could run it during CI. Like I said, I could only run tests for my own new library.
- Setting up continouos delivery to NuGet in Github Actions
- Add StyleCop Analyzer
- Add [SourceLink](https://github.com/dotnet/sourcelink)
- Return `List<T>` instead of `IEnumerable<T>`s. Again, it's just better to return already evaluated data types, rather than lazy ones.
